WalletWasabi.Fluent/ViewModels/CoinSelection/SelectCoinsDialogViewModel.cs

```diff
@@ -0,0 +1,161 @@
+	[AutoNotify] private ReactiveCommand<Unit, Unit> _selectAllPrivateCoinsCommand = ReactiveCommand.Create(() => { });
+	[AutoNotify] private IObservable<Money> _selectedAmount = Observable.Return(Money.Zero);
+	[AutoNotify] private IObservable<int> _selectedCount = Observable.Return(0);
+	[AutoNotify] private ReactiveCommand<Unit, Unit> _selectPredefinedCoinsCommand = ReactiveCommand.Create(() => { });
+	[AutoNotify] private IObservable<string> _summaryText = Observable.Return("");
+
+	public SelectCoinsDialogViewModel(WalletViewModel walletViewModel, TransactionInfo transactionInfo, IObservable<Unit> balanceChanged, IEnumerable<SmartCoin> usedCoins)
+	{
+		_walletViewModel = walletViewModel;
+		_balanceChanged = balanceChanged;
+		_usedCoins = usedCoins;
+		TargetAmount = transactionInfo.MinimumRequiredAmount == Money.Zero
+			? transactionInfo.Amount
+			: transactionInfo.MinimumRequiredAmount;
+
+		_transactionLabels = transactionInfo.UserLabels;
+
+		SetupCancel(enableCancel: false, enableCancelOnEscape: true, enableCancelOnPressed: false);
+		EnableBack = true;
+		NextCommand = ReactiveCommand.Create(() => new List<SelectableCoin>());
+	}
+
+	public Money TargetAmount { get; }
+
+	private new ReactiveCommand<Unit, List<SelectableCoin>> NextCommand { get; set; }
+
+	[SuppressMessage("Reliability", "CA2000:Dispose objects before losing scope", Justification = "Objects use DisposeWith")]
+	protected override void OnNavigatedTo(bool isInHistory, CompositeDisposable disposables)
+	{
+		var coinChanges = _balanceChanged
+			.StartWith()
+			.ObserveOn(RxApp.MainThreadScheduler)
+			.SelectMany(_ => GetCoins())
+			.ToObservableChangeSet(x => x.OutPoint)
+			.TransformWithInlineUpdate(smartCoin => new SelectableCoin(smartCoin), (selectableCoin, smartCoin) => selectableCoin.Coin = smartCoin)
+			.Replay();
+
+		coinChanges.Connect().DisposeWith(disposables);
+
+		coinChanges
+			.Bind(out var coinCollection)
+			.Subscribe()
+			.DisposeWith(disposables);
+
+		var allCoins = coinChanges
+			.AutoRefresh(x => x.IsSelected)
+			.ToCollection()
+			.Throttle(TimeSpan.FromMilliseconds(100), RxApp.MainThreadScheduler);
+
+		var selectedCoins = allCoins
+			.Select(items => items.Where(t => t.IsSelected).ToList());
+
+		EnoughSelected = selectedCoins.Select(coins => coins.Sum(x => x.Amount) >= TargetAmount);
+
+		IsSelectionBadlyChosen = allCoins.Select(coins => IsSelectionBadForPrivacy(coins, _transactionLabels)).ReplayLastActive();
+
+		SelectedAmount = selectedCoins.Select(Sum);
+
+		RemainingAmount = SelectedAmount.Select(money => Money.Max(TargetAmount - money, Money.Zero));
+
+		SelectedCount = selectedCoins.Select(models => models.Count);
+
+		NextCommand = ReactiveCommand.CreateFromObservable(() => selectedCoins, EnoughSelected);
+		NextCommand.Subscribe(models => Close(DialogResultKind.Normal, _walletViewModel.Wallet.Coins.Where(x => models.Any(coin => coin.OutPoint == x.OutPoint))));
```

Comment:
It is not clear to me why `NextCommand` needs to reference `selectedCoins` when `EnoughSelected` already does.

Also, the way we create commands elsewhere is more straightforward with the instantiation referencing the executable action itself (the one in line 104). I would prefer that approach.

Also: line 104 has too many things going on inside of it. I would refactor the lambda into a separate method for clarity, and have the second parameter of `Close()` as a temp variable with a proper name (also for clarity)

Reply:
I've improved the readability of this part. Please, tell me if it's better now. Keep in mind that the `selectedCoins` is and `IObservable<List<ICoin>>`. It doesn't have a "current" value. In order to materialize its last emitted value we need to use `ReactiveCommand.CreateFromObservable`. Otherwise, we should use a `BehaviorSubject` subscribed to the `selectedCoins` sequence. Then, the subject will retain the most recent list of coins, but the code will be more confuse IMHO.